source/fe/fe_system.cc

```diff
@@ -2251,64 +2251,71 @@ FESystem<dim,spacedim>::get_constant_modes () const
+      const auto &base_element = this->base_element(base);
+      const unsigned int multiplicity = this->element_multiplicity(base);
+      const unsigned int n_base_dofs = base_element.dofs_per_cell;
+      const unsigned int n_base_points =
+        base_element.get_generalized_support_points().size();
+      const unsigned int n_base_components = base_element.n_components();
+
+      // FIXME: These temporary vectors are extremely unfortunate. But there
+      // is no easy way around copying all data to make a recursive call...
+      std::vector<double> base_dof_values(n_base_dofs);
+      std::vector<Vector<double> > base_point_values(n_base_points);
+
+      for (unsigned int m = 0; m < multiplicity;
+           ++m, current_vector_component += base_element.n_components())
         {
-          // extract the support points of this base element.
-          // assume that support points are numbered in exactly
-          // the same way as dofs -- this is how the
-          // initialize_unit_support_points() function does it,
-          // at least
-          //
-          // to pick things apart, we could really use a base_to_system_index()
-          // function, but that doesn't exist -- just do it by using
-          // the reverse table -- the amount of work done here is not
-          // worth trying to optimizing this
-          base_support_point_values.resize  (base_element(base).dofs_per_cell);
-          for (unsigned int i=0; i<this->dofs_per_cell; ++i)
-            if (this->system_to_base_index(i).first == std::make_pair(base,m))
-              {
-                const unsigned int base_n_components = base_element(base).n_components();
-                base_support_point_values[this->system_to_base_index(i).second]
-                .reinit (base_n_components, false);
-                for (unsigned int c=0; c<base_n_components; ++c)
-                  base_support_point_values[this->system_to_base_index(i).second][c]
-                    = support_point_values[i][current_vector_component+c];
-              }
+          // populate base_point_values for a recursive call to
+          // convert_generalized_support_point_values_to_dof_values
+          for (unsigned int j = 0; j < base_point_values.size(); ++j)
+            {
+              base_point_values[j].reinit(n_base_components, false);
 
-          // then call the base element to get its version of the
-          // nodal values
-          base_nodal_values.resize (base_element(base).dofs_per_cell);
-          base_element(base).convert_generalized_support_point_values_to_nodal_values
-          (base_support_point_values, base_nodal_values);
+              // we have to extract the correct slice out of the global
+              // vector of values:
+              const auto n = generalized_support_points_index_table[base][j];
+              const VectorSlice<const Vector<double> > slice(
+                point_values[n], current_vector_component, n_base_components);
```

Comment:
This is really just too obscure -- setting up the `VectorSlice` just so you can call `std::copy`. I had to look up what the constructor arguments mean. Can't you just replace all of this with a hand-written loop that cherry-picks the elements from `point_values` and puts them into `base_point_values`? I think that would be a lot simpler to read.

Reply:
I can replace this with raw iterator ranges, but the call to `std::copy` will remain.

Reply:
Incidentally I converted this from the former to the current format because I think the `VectorSlice` greatly increases readability.

Reply:
I don't see that. I don't think it's (i) easier to read, given that I now need to look up what `VectorSlice` does, (ii) faster given that all of the same index translations still have to happen, (iii) shorter, given that I could write down the whole loop that happens in `std::copy` in just half a dozen lines of code, including all of the index translation.

Reply:
The call to std::copy is just four lines (expanded for better readability).